File: halvesAreAlike.java
import java.util.*;
// Appraoch 1- Counting in place
class Solution { // TC O(n) SC O(1)
    public boolean halvesAreAlike(String s) {
        int n = s.length();

        String vowels = "aeiouAEIOU";
        int aVowelCount = 0;
        for (int i=0; i<n/2; i++) {
            if (vowels.indexOf(s.charAt(i)) != -1) {
                aVowelCount++;
            }
        }

        int bVowelCount = 0;
        for (int i =n/2; i<n; i++) {
            if (vowels.indexOf(s.charAt(i)) != -1) {
                bVowelCount++;
            }
        }

        return aVowelCount == bVowelCount;
    }
}

class halvesAreAlike{
    public static void main(String[] args) {
        String x= "book";
        Solution s= new Solution();
        System.out.println(s.halvesAreAlike(x));
    }
}

File: halvesAreAlike.py
# Approach 1- Counting in place 
def halvesAreAlike(s): # TC O(n) // SC O(1)
	vowels="aeiouAEIOU"
	n=len(s)
	aVowel=0
	for i in range(n//2):
	    if s[i] in vowels:
	        aVowel+=1

	bVowel=0
	for i in range(n//2,n):
	    if s[i] in vowels:
	        bVowel+=1

	return aVowel==bVowel

print(halvesAreAlike("book"))